### ISSUE TITLE
Running trains local server behind proxy

### ISSUE THREAD

**Issue by manish2295** [allegroai/clearml#126]
Getting the following error when trying to run mnist.py on local trains server that is behind proxy.

Retrying (Retry(total=239, connect=240, read=240, redirect=240, status=240)) after connection broken by 'ProxyError('Cannot connect to proxy.', NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f5bacb8a048>: Failed to establish a new connection: [Errno 111] Connection refused',))'

**Comment by bmartinn** [comment]
Hi @manish2295 ,
Seems like you pointed right at the problem, somehow the `requests` pacakge cannot connect with the `trains-server`. 
How did you configure the system wide proxy ?
What OS are you working on?

Notice that the `trains-server` needs three different ports, 8080, 8008, 8081.

Specifically the *trains* package is using port 8008 for the initial handshake, can you access the `trains-server` on port 8008 manually ?

**Comment by manish2295** [comment]
I have configure using export http_proxy and https_proxy

I am using Ubuntu 16.04

Specifically the trains package is using port 8008 for the initial handshake, can you access the trains-server on port 8008 manually ? -- I didn't get this.

But, when I try localhost:8008 I am getting the following output:

{"meta":{"id":"cc1f644103d34f2e8dbd153b621a5b7c","trx":"cc1f644103d34f2e8dbd153b621a5b7c","endpoint":{"name":"","requested_version":1.0,"actual_version":null},"result_code":400,"result_subcode":0,"result_msg":"Invalid request path /","error_stack":null},"data":{}}

**Comment by bmartinn** [comment]
Hi @manish2295 seems like you are getting the correct response.

Just making sure here, your trains.conf points to `localhost`, correct? your `~/trains.conf` file  should look something like
```
api {
    api_server: http://localhost:8008
    web_server: http://localhost:8080
    files_server: http://localhost:8081
}
```

If you are running the *trains-server* on another machine, and `localhost:8008` is answering, I assume you have port-forwarding to the the  remote *trains-server*, is this correct?

If this is the case, I strongly recommend configuring your machine (the one with the *trains* python installed) with some host name point to 127.0.0.1, for example in your `/etc/hosts` add:
`127.0.0.1 trains-server`

Reason being, if you have the `files_server: http://localhost:8081` all your artifacts will point to this link e.g. `http://localhost:8081/<project>/<task>/artifact` , this will prevent other users from accessing those links (unless of course they also use port-forwarding like you do)

**Comment by manish2295** [comment]
/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py:1004: InsecureRequestWarning: Unverified HTTPS request is being made to host 'proxy-basic.rd.corpintra.net'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/manish/.local/lib/python3.5/site-packages/trains/task.py", line 277, in init
    not auto_connect_frameworks.get('detect_repository', True)) else True
  File "/home/manish/.local/lib/python3.5/site-packages/trains/task.py", line 1163, in _create_dev_task
    log_to_backend=True,
  File "/home/manish/.local/lib/python3.5/site-packages/trains/task.py", line 111, in __init__
    super(Task, self).__init__(**kwargs)
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_interface/task/task.py", line 90, in __init__
    super(Task, self).__init__(id=task_id, session=session, log=log)
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_interface/base.py", line 129, in __init__
    super(IdObjectBase, self).__init__(session, log, **kwargs)
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_interface/base.py", line 34, in __init__
    self._session = session or self._get_default_session()
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_interface/base.py", line 103, in _get_default_session
    secret_key=ENV_SECRET_KEY.get(),
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_api/session/session.py", line 144, in __init__
    self.refresh_token()
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_api/session/token_manager.py", line 95, in refresh_token
    self._set_token(self._do_refresh_token(self.__token, exp=self.req_token_expiration_sec))
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_api/session/session.py", line 547, in _do_refresh_token
    six.reraise(*sys.exc_info())
  File "/usr/local/lib/python3.5/dist-packages/six.py", line 703, in reraise
    raise value
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_api/session/session.py", line 540, in _do_refresh_token
    res.status_code, self.host, msg
trains.backend_api.session.session.LoginError: Failed getting token (error 401 from http://dataanalytics2:8008): Authorization Required

Getting the above error when I do Task.init()

and when I run http://dataanalytics2:8008/ in the browser I get the following output:

"meta":{"id":"d652587b0bcf4133b0368bf379f8f52e","trx":"d652587b0bcf4133b0368bf379f8f52e","endpoint":{"name":"","requested_version":1.0,"actual_version":null},"result_code":400,"result_subcode":0,"result_msg":"Invalid request path /","error_stack":null},"data":{}}

**Comment by bmartinn** [comment]
Ohhh @manish2295 now I see the problem,
This is not a proxy issue, it's SSL certificate issue.
For a lot of companies the firewall "wants" to look into the traffic the https connections, so essentially it does man-in-the-middle and replaces the security certificate with an unverified one, this is happening on your proxy connection 
> error "Unverified HTTPS request is being made to host 'proxy-basic.rd.corpintra.net'. " .

The quickest solution is adding not veterinary the SSL certificate, just add in the `~/trains.conf` :
```
api { 
    verify_certificate = False 
} 
```

Basically the same issue as this one #7

**Comment by manish2295** [comment]
Hi @bmartinn,
after making the above change in trains.conf file, I got the following error:

2020-04-26 17:48:26,775 - trains - WARNING - InsecureRequestWarning: Certificate verification is disabled! Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/manish/.local/lib/python3.5/site-packages/trains/task.py", line 277, in init
    not auto_connect_frameworks.get('detect_repository', True)) else True
  File "/home/manish/.local/lib/python3.5/site-packages/trains/task.py", line 1163, in _create_dev_task
    log_to_backend=True,
  File "/home/manish/.local/lib/python3.5/site-packages/trains/task.py", line 111, in __init__
    super(Task, self).__init__(**kwargs)
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_interface/task/task.py", line 90, in __init__
    super(Task, self).__init__(id=task_id, session=session, log=log)
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_interface/base.py", line 129, in __init__
    super(IdObjectBase, self).__init__(session, log, **kwargs)
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_interface/base.py", line 34, in __init__
    self._session = session or self._get_default_session()
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_interface/base.py", line 103, in _get_default_session
    secret_key=ENV_SECRET_KEY.get(),
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_api/session/session.py", line 144, in __init__
    self.refresh_token()
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_api/session/token_manager.py", line 95, in refresh_token
    self._set_token(self._do_refresh_token(self.__token, exp=self.req_token_expiration_sec))
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_api/session/session.py", line 547, in _do_refresh_token
    six.reraise(*sys.exc_info())
  File "/usr/local/lib/python3.5/dist-packages/six.py", line 703, in reraise
    raise value
  File "/home/manish/.local/lib/python3.5/site-packages/trains/backend_api/session/session.py", line 540, in _do_refresh_token
    res.status_code, self.host, msg
trains.backend_api.session.session.LoginError: Failed getting token (error 401 from http://dataanalytics2:8008): Authorization Required

**Comment by bmartinn** [comment]
@manish2295 my bad, it is not the SSL certificate verification, it's actually the proxy.
We have noticed that on some proxies (like on AWS) the proxy does not forward the header of the GET request, that header stores the login credentials, so the *trains-server* ends up getting a login request without the credentials (as the proxy stripped them off the packet)

This behavior does not happen on https proxy=ing (don't ask me why ...), 
so if you use  `https://dataanalytics2:8008` it should work
Notice that if you have a self signed certificate, just leave the `verify_certificate = False` , and it should work.
